modulus/datapipes/climate/climate.py

```diff
@@ -0,0 +1,829 @@
+
+            if not mean_stat_file.exists():
+                raise IOError(f"Mean statistics file {mean_stat_file} not found")
+            if not std_stat_file.exists():
+                raise IOError(f"Std statistics file {std_stat_file} not found")
+
+            # has shape [1, C, 1, 1]
+            self.mu = np.load(str(mean_stat_file))[:, self.channels]
+            # has shape [1, C, 1, 1]
+            self.sd = np.load(str(std_stat_file))[:, self.channels]
+        elif set(self.stats_files) == {
+            "norm",
+        }:  # use dict formatted file with named variables
+            norm_stat_file = self.stats_files["norm"]
+            if not norm_stat_file.exists():
+                raise IOError(f"Statistics file {norm_stat_file} not found")
+
+            norm = np.load(str(norm_stat_file), allow_pickle=True).item()
+            mu = np.array([norm[var]["mean"] for var in self.variables])
+            self.mu = mu.reshape((1, len(mu), 1, 1))
+            sd = np.array([norm[var]["std"] for var in self.variables])
+            self.sd = sd.reshape((1, len(sd), 1, 1))
+        else:
+            raise ValueError(("Invalid statistics file specification"))
+
+        if not self.mu.shape == self.sd.shape == (1, len(self.channels), 1, 1):
+            raise ValueError("Error, normalisation arrays have wrong shape")
+
+
+class ClimateDatapipe(Datapipe):
+    """
+    A Climate DALI data pipeline. This pipeline loads data from
+    HDF5/NetCDF4 files, which can include latitude, longitude, cosine of the
+    solar zenith angle, geopotential, and land-sea mask if specified. Additionally,
+    it normalizes
+    the data if a statistics file is provided. The pipeline returns a dictionary
+    with the following structure, where {name} indicates the name of the data
+    source provided:
+
+    - `state_seq-{name}`: Tensors of shape
+      (batch_size, num_steps, num_channels, height, width).
+      This sequence is drawn from the data file and normalized if a
+      statistics file is provided.
+    - `timestamps-{name}`: Tensors of shape (batch_size, num_steps), containing
+      timestamps for each timestep in the sequence.
+    - `land_sea_mask`: Tensor of shape (batch_size, 1, height, width),
+      containing the land sea mask if a path to a land sea mask file is
+      provided.
+    - `geopotential`: Tensor of shape (batch_size, 1, height, width), containing
+      geopotential if a path to a geopotential file is provided.
+    - `latlon`: Tensor of shape (batch_size, 2, height, width), containing
+      latitude and longitude meshgrid if specified.
+    - `cos_latlon`: Tensor of shape (batch_size, 3, height, width), containing
+      `[cos(lat), sin(lon), cos(lon)]` if specified. This is required by many
+      neural climate models.
+    - `cos_zenith-{name}`: Tensors of shape (batch_size, num_steps, 1, height, width),
+      containing the cosine of the solar zenith angle if specified.
```

Comment:
I don't think I support hard-coding these names into the datapipe because it restricts/confines exactly what inputs users need to use with this datapipe. If I wanted to bring in another variable then do I have to make my own datapipe, even if everything else is the same? I think we should think of a solution that adds variables in a more flexible way.

Reply:
Yes, these were ported over from the `ClimateHDF5Datapipe` in `experimental`. Having a more flexible system, where more auxiliary variables can be added by the user, would be nice. At the same time I think it's good to offer some of the most common ones out of the box.

Reply:
A dict-based interface to adding auxiliary variables is now implemented in e3a916f6a3b9bb4984f13b67fc53a41fe884624b.